FWCore/Services/plugins/InitRootHandlers.cc

```diff
@@ -346,7 +377,55 @@ namespace {
       return full_write(2, text);
     }
 
+    void sig_pause_for_stacktrace(int sig, siginfo_t*, void*) {
+      char buff[128] = "\nModule: ";
+ 
+      if (edm::CurrentModuleOnThread::getCurrentModuleOnThread() != nullptr) {
+        strlcat(buff, edm::CurrentModuleOnThread::getCurrentModuleOnThread()->moduleDescription()->moduleName().c_str(), 128);
+        full_cerr_write(buff);
+      }
+
+      struct timespec timeout = { edm::service::InitRootHandlers::stackTracePause(), 0 };
+#if defined(__linux__)
+      sigset_t sigset;
+      sigemptyset(&sigset);
+      sigaddset(&sigset, SIGCONT);
+      sigtimedwait(&sigset, NULL, &timeout);
+#else
+      nanosleep(&timeout); // macOS doesn't have sigtimedwait()
+#endif
+    }
+
     void sig_dostack_then_abort(int sig, siginfo_t*, void*) {
+      char buff[128] = "\nModule: ";
+
+      auto&& tids = edm::service::InitRootHandlers::threadIDs();
+
+      const auto self = pthread_self();
+#ifdef PAUSE_SIGNAL
+      if (edm::service::InitRootHandlers::stackTracePause() > 0 && tids.size() > 1) {
+        sigset_t pausesigset;
+        sigemptyset(&pausesigset);
+        sigaddset(&pausesigset,PAUSE_SIGNAL);
+        sigprocmask(SIG_UNBLOCK,&pausesigset,0);
+
+        int notified = 0;
+        for (auto id : tids) {
+          if (self != id) {
+            if (pthread_kill(id, PAUSE_SIGNAL) == 0) ++notified;
+          }
+        }
+      }
+#endif
+
+      // only access current module thread local if this is a CMSSW thread; assumes that any TBB
+      // thread will have access to CMSSW thread locals.
+      if (tids.count(self) > 0 && edm::CurrentModuleOnThread::getCurrentModuleOnThread() != nullptr)
```

Comment:
> On Aug 9, 2016, at 11:17, Brian Bockelman <notifications@github.com> wrote:
> 
> > +      // only access current module thread local if this is a CMSSW thread; assumes that any TBB
> > +      // thread will have access to CMSSW thread locals.
> > +      if (tids.count(self) > 0 && edm::CurrentModuleOnThread::getCurrentModuleOnThread() != nullptr)
> 
> Did you inspect getCurrentModuleOnThread to verify this is a safe function to call? If so, can you add a comment along these lines?

If

static thread_local ModuleCallingContext const* currentModuleOnThread_;

has been allocated, then everything else is safe.  Relying on a thread local in a signal handler, though, is the thing I'm most nervous about.

-dan